File: FacialMotionCapture/FacialMotionCapture/UDPSocket.swift
//
//  UDPSocket.swift
//  FacialMotionCapture
//
//  Created by Jihyun Lee on 5/5/20.
//  Copyright © 2020 Jihyun Lee. All rights reserved.
//

import Foundation

@_silgen_name("udpsocket_open") func c_udpsocket_open() -> Int32
@_silgen_name("udpsocket_close") func c_udpsocket_close(_ socketId:Int32) -> Int32
@_silgen_name("udpsocket_enable_broadcast") func c_udpsocket_enable_broadcast(_ socketId:Int32) -> Void
@_silgen_name("udpsocket_send") func c_udpsocket_send(_ socketId:Int32,buff:UnsafePointer<UInt8>,ip:UnsafePointer<Int8>,port:Int32) -> Int32
@_silgen_name("udpsocket_broadcast") func c_udpsocket_broadcast(_ socketId:Int32,buff:UnsafePointer<UInt8>,port:Int32) -> Int32

@objc open class UDPBroadcaster: NSObject {
    
    var ip: String
    var port: Int
    var socketId: Int32?
    var ipAddresses: [String] = []
    
    init(port: Int) {
        self.ip = ""
        self.port = port
        
        let socketId:Int32 = c_udpsocket_open()
        if socketId > 0 {
            self.socketId = socketId
        }
    }
    
    init(ip :String, port :Int) {
        self.ip = ip
        self.port = port

        let socketId:Int32 = c_udpsocket_open()
        if socketId > 0 {
            self.socketId = socketId
        }
    }
    
    public func setIpAddresses(array: [String]) {
        self.ipAddresses = array
    }

    public func enableBroadcast() {
        if let socketId:Int32 = self.socketId {
            c_udpsocket_enable_broadcast(socketId)
        }
    }
    
    public func send(_ string: String)->(Bool,String) {
        if string.count <= 9216 {
            
            if let socketId:Int32 = self.socketId {
                let ret:Int32 = c_udpsocket_send(socketId, buff: string, ip:self.ip, port: Int32(self.port))
                if ret == Int32(strlen(string)) {
                    return (true,"send success")
                } else {
                    return (false,"send error")
                }
            } else{
                return (false,"socket not open")
            }
        }
        
        print("packet is too large. Must be smaller than 9200 bytes")
        return (false, "packet is too large")
    }
    
    public func sendToList(_ string: String) {
        for addr in self.ipAddresses {
            if string.count <= 9216 {
                if let socketId:Int32 = self.socketId {
                    _ = c_udpsocket_send(socketId, buff: string, ip:addr, port: Int32(self.port))
                }
            }
        }
    }

    public func broadcast(_ string: String)->(Bool,String) {
        if string.count <= 9216 {
            
            if let socketId:Int32 = self.socketId {
                let ret:Int32 = c_udpsocket_broadcast(socketId, buff: string, port: Int32(self.port))
                if ret == Int32(strlen(string)) {
                    return (true,"send success")
                } else {
                    return (false,"send error")
                }
            } else{
                return (false,"socket not open")
            }
        }
        
        print("packet is too large. Must be smaller than 9200 bytes")
        return (false, "packet is too large")
    }

    deinit {
        if let socketId:Int32=self.socketId{
            _ = c_udpsocket_close(socketId)
            self.socketId = nil
        }
    }
}

File: FacialMotionCapture/FacialMotionCapture/UDPSocket.c
//
//  UDPSocket.c
//  FacialMotionCapture
//
//  Created by Jihyun Lee on 5/5/20.
//  Copyright © 2020 Jihyun Lee. All rights reserved.
//

#include <stdio.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <string.h>
#include <unistd.h>
#include <netdb.h>

int udpsocket_open()
{
    int socketId = socket(AF_INET, SOCK_DGRAM, 0);
    if (socketId < 0) {
        printf("udpsocket_broadcast_client Error: Could not open socket.");
        return -1;
    }
    
    int reuseon   = 1;
    setsockopt(socketId, SOL_SOCKET, SO_REUSEADDR, &reuseon, sizeof(reuseon));

    return socketId;
}

int udpsocket_close(int socketId)
{
    return close(socketId);
}

void udpsocket_enable_broadcast(int socketId)
{
    // set socket options enable broadcast
    int broadcastEnable = 1;
    int ret = setsockopt(socketId, SOL_SOCKET, SO_BROADCAST, &broadcastEnable, sizeof(broadcastEnable));
    if (ret < 0) {
        printf("udpsocket_broadcast_client Error: Could not open set socket to broadcast mode");
    }
}

int udpsocket_send(int socketId, char *msg, char *ip, int port)
{
    struct sockaddr_in addr;
    memset(&addr, 0x0, sizeof(struct sockaddr_in));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(port);
    addr.sin_addr.s_addr = inet_addr(ip);

    int ret = (int) sendto(socketId, msg, strlen(msg), 0, (struct sockaddr *)&addr, sizeof(addr));
    if (ret < -1) {
        printf("Error: Could not send message.");
    }
    else {
//        printf("sent !!!!! %d\n", ret);
    }
    return ret;
}

int udpsocket_broadcast(int socketId, char *msg, int port)
{
//    printf("udpsocket_broadcast: (%lu bytes)\n%s\n", strlen(msg), msg);

    struct sockaddr_in broadcastAddr;
    memset(&broadcastAddr, 0, sizeof(broadcastAddr));
    broadcastAddr.sin_family = AF_INET;
    broadcastAddr.sin_port = htons(port);
    broadcastAddr.sin_addr.s_addr = htonl(INADDR_BROADCAST);
    
    int ret = (int) sendto(socketId, msg, strlen(msg), 0, (struct sockaddr*)&broadcastAddr, sizeof(broadcastAddr));
    if (ret < -1) {
        printf("Error: Could not open send broadcast.");
    }
    else {
//        printf("broadcasted !!!!! %d\n", ret);
    }

    return ret;
}
